### PR TITLE
Simplify on-chain fee management 

### PR DESCRIPTION
**Pull request by pm47**
Move away from the "block target" approach.

Get rid of the `FeeEstimator` abstraction and use an `AtomicReference` to store and update the current feerates, similar to the block count.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala [hunk 1]
```diff
@@ -62,14 +47,18 @@ case class FeerateTolerance(ratioLow: Double, ratioHigh: Double, anchorOutputMax
   }
 }
 
-case class OnChainFeeConf(feeTargets: FeeTargets, feeEstimator: FeeEstimator, spendAnchorWithoutHtlcs: Boolean, closeOnOfflineMismatch: Boolean, updateFeeMinDiffRatio: Double, private val defaultFeerateTolerance: FeerateTolerance, private val perNodeFeerateTolerance: Map[PublicKey, FeerateTolerance]) {
+case class OnChainFeeConf(feerates: AtomicReference[FeeratesPerKw], safeUtxosThreshold: Int, spendAnchorWithoutHtlcs: Boolean, closeOnOfflineMismatch: Boolean, updateFeeMinDiffRatio: Double, private val defaultFeerateTolerance: FeerateTolerance, private val perNodeFeerateTolerance: Map[PublicKey, FeerateTolerance]) {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: Putting `feerates` is not very pure (because the case class is now mutable), but it simplify greatly the code by containing everything related to feerates in `OnChainFeeConf`.

Unfortunately we can't make `feerates` private in order to make it testable.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala [hunk 2]
```diff
@@ -62,14 +47,18 @@ case class FeerateTolerance(ratioLow: Double, ratioHigh: Double, anchorOutputMax
...
+
+  def currentFeerates: FeeratesPerKw = feerates.get()
 
   def feerateToleranceFor(nodeId: PublicKey): FeerateTolerance = perNodeFeerateTolerance.getOrElse(nodeId, defaultFeerateTolerance)
 
   /** To avoid spamming our peers with fee updates every time there's a small variation, we only update the fee when the difference exceeds a given ratio. */
   def shouldUpdateFee(currentFeeratePerKw: FeeratePerKw, nextFeeratePerKw: FeeratePerKw): Boolean =
     currentFeeratePerKw.toLong == 0 || Math.abs((currentFeeratePerKw.toLong - nextFeeratePerKw.toLong).toDouble / currentFeeratePerKw.toLong) > updateFeeMinDiffRatio
 
+  def getFundingFeerate(): FeeratePerKw = currentFeerates.blocks_6
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:60]: Note how we now have `getFundingFeerate`/`getCommitmentFeerate`/`getClosingFeerate` in this class, which are used most of the time. Only in advanced scenarii (revoked, fee bumping) does code directly use `currentFeerates`.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala [hunk 3]
```diff
@@ -79,11 +68,10 @@ case class OnChainFeeConf(feeTargets: FeeTargets, feeEstimator: FeeEstimator, sp
    * @param channelType         channel type
    * @param currentFeerates_opt if provided, will be used to compute the most up-to-date network fee, otherwise we rely on the fee estimator
    */
-  def getCommitmentFeerate(remoteNodeId: PublicKey, channelType: SupportedChannelType, channelCapacity: Satoshi, currentFeerates_opt: Option[CurrentFeerates]): FeeratePerKw = {
-    val (networkFeerate, networkMinFee) = currentFeerates_opt match {
-      case Some(currentFeerates) => (currentFeerates.feeratesPerKw.feePerBlock(feeTargets.commitmentBlockTarget), currentFeerates.feeratesPerKw.mempoolMinFee)
-      case None => (feeEstimator.getFeeratePerKw(feeTargets.commitmentBlockTarget), feeEstimator.getMempoolMinFeeratePerKw())
-    }
+  def getCommitmentFeerate(remoteNodeId: PublicKey, channelType: SupportedChannelType, channelCapacity: Satoshi): FeeratePerKw = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:71]: I removed the `currentFeerates_opt: Option[CurrentFeerates]`. The reason for this argument was that when the channel received a `CurrentFeerateEvent` and calls `getCommitmentFeerate`, it has to be sure that the latest feerate will be used.

I think it's okay to make the assumption that the atomic reference is updated before  `CurrentFeerate` is emitted, which is already the case in our implementation.

A drawback is that there will be a herd effect every 10min, where all channels will attempt to read the atomic reference at the same time. Is it a big deal? I'm not sure what the locking contraints are. We can work around that by using the same trick as `ProcessCurrentBlockHeight`, which was introduced for that purpose.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 4]
```diff
@@ -634,16 +634,16 @@ object Helpers {
         }
         // NB: we choose a minimum fee that ensures the tx will easily propagate while allowing low fees since we can
         // always use CPFP to speed up confirmation if necessary.
-        val closingFeerates = ClosingFeerates(preferredFeerate, preferredFeerate.min(feeEstimator.getFeeratePerKw(1008)), preferredFeerate * 2)
+        val closingFeerates = ClosingFeerates(preferredFeerate, preferredFeerate.min(onChainFeeConf.currentFeerates.mempoolMinFee), preferredFeerate * 2)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:637]: Using 1008 blocks instead of `mempoolMinFee` looks like a bug.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 5]
```diff
@@ -748,7 +748,7 @@ object Helpers {
         val spendAnchors = htlcTxs.nonEmpty || onChainFeeConf.spendAnchorWithoutHtlcs
         val claimAnchorTxs: List[ClaimAnchorOutputTx] = if (spendAnchors) {
           // If we don't have pending HTLCs, we don't have funds at risk, so we can aim for a slower confirmation.
-          val confirmCommitBefore = htlcTxs.values.flatten.map(htlcTx => htlcTx.confirmBefore).minOption.getOrElse(currentBlockHeight + onChainFeeConf.feeTargets.commitmentWithoutHtlcsBlockTarget)
+          val confirmCommitBefore = htlcTxs.values.flatten.map(htlcTx => htlcTx.confirmBefore).minOption.getOrElse(currentBlockHeight + 1008) // TODO: setting a block target may not make sense when there is no urgency
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:751]: There is something incoherent with using a fixed target (even far in the future) for txs that are not urgent. In some full mempool scenarii we may find ourselves agressively bidding to get transactions confirmed when there is no real emergency.

Should we make the `confirmBefore` optional and reserve it to cases where there is a hard limit in terms of block height?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 6]
```diff
@@ -820,9 +820,9 @@ object Helpers {
        * NB: with anchor outputs, it's possible to have transactions that spend *many* HTLC outputs at once, but we're not
        * doing that because it introduces a lot of subtle edge cases.
        */
-      def claimHtlcDelayedOutput(localCommitPublished: LocalCommitPublished, keyManager: ChannelKeyManager, commitment: FullCommitment, tx: Transaction, feeEstimator: FeeEstimator, feeTargets: FeeTargets, finalScriptPubKey: ByteVector)(implicit log: LoggingAdapter): (LocalCommitPublished, Option[HtlcDelayedTx]) = {
+      def claimHtlcDelayedOutput(localCommitPublished: LocalCommitPublished, keyManager: ChannelKeyManager, commitment: FullCommitment, tx: Transaction, onChainFeeConf: OnChainFeeConf, finalScriptPubKey: ByteVector)(implicit log: LoggingAdapter): (LocalCommitPublished, Option[HtlcDelayedTx]) = {
         if (isHtlcSuccess(tx, localCommitPublished) || isHtlcTimeout(tx, localCommitPublished)) {
-          val feeratePerKwDelayed = feeEstimator.getFeeratePerKw(feeTargets.claimMainBlockTarget)
+          val feeratePerKwDelayed = onChainFeeConf.getClosingFeerate()
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:825]: I lumped all "non-competitive" cases (mutual close, claim main, claim htlc delayed) in the same `getClosingFeerate()` bucket. I don't see why we should separate them.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala [hunk 7]
```diff
@@ -18,25 +18,10 @@ package fr.acinq.eclair.blockchain.fee
...
 
-trait FeeEstimator {
-  // @formatter:off
-  def getFeeratePerKb(target: Int): FeeratePerKB
-  def getFeeratePerKw(target: Int): FeeratePerKw
-  def getMempoolMinFeeratePerKw(): FeeratePerKw
-  // @formatter:on
-}
-
-case class FeeTargets(fundingBlockTarget: Int, commitmentBlockTarget: Int, commitmentWithoutHtlcsBlockTarget: Int, mutualCloseBlockTarget: Int, claimMainBlockTarget: Int, safeUtxosThreshold: Int) {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:33]: We may still want to keep a slow/fast preference for funding and closing txs, that would fit in `OnChainFeeConf`. WDYT?

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:33]: Yes definitely, I think the node operator should be able to define in `eclair.conf` something like the following:

```
confirmation-target {
      // target for the funding transaction
      funding = slow/medium/fast
      // target for the commitment transaction (used in force-close scenario) *do not change this unless you know what you are doing*
      commitment = slow/medium/fast
      // target for the commitment transaction when we have no htlcs to claim (used in force-close scenario) *do not change this unless you know what you are doing*
      commitment-without-htlcs = slow/medium/fast
      // target for the mutual close transaction
      mutual-close = slow/medium/fast
      // target for the claim main transaction (tx that spends main channel output back to wallet)
      claim-main = slow/medium/fast
    }
```

Those were previously a block target, they should now be an enum `slow`, `medium` or `fast`?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: It's a bit frustrating though...everywhere we use an `OnChainFeeConf`, we actually have access to the whole `NodeParams` field, so we could keep `OnChainFeeConf` pure and pass the `feerates` separately, I'd feel more comfortable with that.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: If this contained a new kind of `FeeTargets` where each kind of transaction (funding, closing, commit, commit without htlcs) had a `slow/medium/fast` target, we could just expose `getFundingFeerate(feerates: AtomicReference[FeeratesPerKw]): FeeratePerKw` (and same for closing/commit).

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 5 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:751]: > Should we make the `confirmBefore` optional and reserve it to cases where there is a hard limit in terms of block height?

I think that's a good idea indeed and probably what I should have done when I added the `commitmentWithoutHtlcsBlockTarget`. If we have no HTLC in the commitment, you're right that there is no specific block before which we must confirm, we should just use the priority (slow/medium/fast) that the node operator set for this case.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 6 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:825]: Agreed, that makes sense! Then we only need the following buckets in `eclair.conf`:

- funding (slow/medium/fast)
- commitment-without-anchors (slow/medium/fast)
- commitment-with-anchors (slow/medium/fast)
- close (slow/medium/fast)

I think we don't need the commitment-without-htlc case anymore, this is handled by separating the anchor/non-anchor cases.

In the anchor case, we will very likely use a defaut of `slow` for the commitment, and the feerate we use when force-closing depends on whether there are HTLCs at risk and when they timeout.

In the non-anchor case, the default should likely be `fast` since we cannot CPFP.

Do you think that makes sense?

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:825]: > Do you think that makes sense?

For funding and close it definitely makes sense to offer configurable priority. But for commitments it should be entirely managed automatically by eclair I think, with a fallback to the regular "close" setting when there are no competing transactions.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: I started that way, it does make `NodeParams` nicer with a perfect symmetry between the atomic references for `blockCount` and `feerates`. However, it made the code in `Commitments` and `Helpers` quite tedious, e.g. I had to pass two args like so:
```scala
def canSendAdd(amount: MilliSatoshi, params: ChannelParams, changes: CommitmentChanges, feeConf: OnChainFeeConf, feerates: FeeratesPerKw)
```

Also, the current change doesn't make the code "less pure" because we were already storing the `FeeEstimator` inside `OnChainFeeConf`.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: > Also, the current change doesn't make the code "less pure" because we were already storing the FeeEstimator inside OnChainFeeConf.

Huh, that's true! But we now have an opportunity to make it more pure :smile: 

I think it makes more sense to have a `XXXConf` class be a really pure class, and other dynamic objects be passed separately: even for unit tests, it's simpler because the thing we'll mock is only the `AtomicRef`, right?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 6 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:825]: > People will have different time preference here depending on whether they have big pockets and want to optimize for fees or not so much liquidity and want to move it quickly.

I agree, but why would this setting be different from the `closing` one?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:50]: > Huh, that's true! But we now have an opportunity to make it more pure 😄

Fair enough! will give it a try

> for unit tests, it's simpler because the thing we'll mock is only the AtomicRef, right?

It will be the same I think, we don't use mocking, simply directly set the atomic ref value

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 6 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala:825]: > I agree, but why would this setting be different from the closing one?

Probably not, good idea, we can treat this as a closing!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 4 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Done in 133264741 (look! an all-digit commit hash).

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala (hunk 5 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: Done in respectively 133264741 and 785d30ba8.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala (hunk 7 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala:33]: Done in 133264741.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala]: Done in b8f334e7a.

**Comment by pm47** [comment]
Rebased on master.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala [hunk 8]
```diff
@@ -75,28 +75,31 @@ object ReplaceableTxPublisher {
...
+        val remainingBlocks = confirmBefore - currentBlockHeight
+        if (hasEnoughSafeUtxos) {
+          remainingBlocks match {
+            // If our target is still very far in the future, no need to rush
+            case t if t >= 36 => feerates.slow
+            // However, if we get closer to the target, we start being more aggressive
+            case t if t >= 12 => feerates.medium
+            case t if t >= 2 => feerates.fast
+            case _ => feerates.fastest
+          }
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala:91]: This changes the dynamic of the fee bumping and is somewhat tailored to mempool-style fee estimation.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/BitcoinCoreFeeProvider.scala [hunk 9]
```diff
@@ -50,22 +50,14 @@ case class BitcoinCoreFeeProvider(rpcClient: BitcoinJsonRPCClient, defaultFeerat
...
-    blocks_6 = if (blocks_6.feerate > 0.sat) blocks_6 else defaultFeerates.blocks_6,
-    blocks_12 = if (blocks_12.feerate > 0.sat) blocks_12 else defaultFeerates.blocks_12,
-    blocks_36 = if (blocks_36.feerate > 0.sat) blocks_36 else defaultFeerates.blocks_36,
-    blocks_72 = if (blocks_72.feerate > 0.sat) blocks_72 else defaultFeerates.blocks_72,
-    blocks_144 = if (blocks_144.feerate > 0.sat) blocks_144 else defaultFeerates.blocks_144,
-    blocks_1008 = if (blocks_1008.feerate > 0.sat) blocks_1008 else defaultFeerates.blocks_1008)
+    minimum = if (mempoolMinFee.feerate > 0.sat) mempoolMinFee else defaultFeerates.minimum,
+    fastest = if (block_1.feerate > 0.sat) block_1 else defaultFeerates.fastest,
+    fast = if (blocks_2.feerate > 0.sat) blocks_2 else defaultFeerates.fast,
+    medium = if (blocks_12.feerate > 0.sat) blocks_12 else defaultFeerates.medium,
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/BitcoinCoreFeeProvider.scala:59]: Note how we use 12 blocks for the medium setting, which would correspond to the `hourFee` mempool fee estimate. It takes into account the bitcoin core fee estimator is conservative.

**Comment by pm47** [comment]
One thing left to address are the default feerates which I would really like to remove. I think there are two reasons for them to exist:
- as constant values for regtest (but we can directly use hardcoded values with a `ConstantFeeProvider`
- to deal with `-1` values returned by `estimatesmartfee` in some corner cases (e.g. restart of bitcoin core)

Not sure exactly how to proceed though!

File: eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalStateSpec.scala [hunk 10]
```diff
@@ -3422,16 +3431,15 @@ class NormalStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLike with
...
       val claimMain = alice2blockchain.expectMsgType[PublishFinalTx]
       assert(alice2blockchain.expectMsgType[WatchTxConfirmed].txId === aliceCommitTx.txid)
       assert(alice2blockchain.expectMsgType[WatchTxConfirmed].txId === claimMain.tx.txid)
       alice2blockchain.expectNoMessage(1 second)
     } else {
       val localAnchor = alice2blockchain.expectMsgType[PublishReplaceableTx]
-      // When there are no pending HTLCs, there is no rush to get the commit tx confirmed
-      assert(localAnchor.txInfo.confirmBefore === currentBlockHeight + blockTargets.commitmentWithoutHtlcsBlockTarget)
+      // When there are no pending HTLCs, there is no absolute deadline to get the commit tx confirmed, we use priority
+      assert(localAnchor.txInfo.confirmationTarget.isInstanceOf[ConfirmationTarget.Priority])
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalStateSpec.scala:3442]: nit: we can even be more accurate:

```suggestion
      assert(localAnchor.txInfo.confirmationTarget == ConfirmationTarget.Priority(ConfirmationPriority.Medium))
```

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala [hunk 11]
```diff
@@ -220,23 +221,33 @@ private class TxPublisher(nodeParams: NodeParams, factory: TxPublisher.ChildFact
...
             }
+
+            (currentConfirmationTarget, proposedConfirmationTarget) match {
+              case (ConfirmationTarget.Absolute(currentConfirmBefore), ConfirmationTarget.Absolute(proposedConfirmBefore)) if proposedConfirmBefore < currentConfirmBefore =>
+                // The proposed block target is closer than what it was
+                updateConfirmationTarget()
+              case (_: ConfirmationTarget.Priority, ConfirmationTarget.Absolute(proposedConfirmBefore)) =>
+                // Switch from relative priority mode to absolute blockheight mode
+                updateConfirmationTarget()
+              case _ =>
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala:247]: It could also make sense to handle the case where both targets are using `ConfirmationTarget.Priority` and the new priority is higher (e.g. the node operator has unconfirmed commit txs without htlcs, wants to reclaim their pending liquidity and updates their `closing` priority in `eclair.conf` from `slow` to `medium` or `fast`).

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala [hunk 12]
```diff
@@ -216,23 +223,37 @@ private class ReplaceableTxPublisher(nodeParams: NodeParams,
...
+              log.debug("{} priority {}: bumping fees", cmd.desc, priority)
+              Some(currentFeerate)
+            } else {
+              log.debug("{} priority {}: no need to bump fees", cmd.desc, priority)
+              None
+            }
         }
+
+
+
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala:250]: nit: remove?

**Comment by t-bast** [comment]
> Not sure exactly how to proceed though!

I don't know what we can do honestly, since we want to handle `bitcoind` restarts...I don't think we want to persist the latest fees, it's better to avoid a frequent DB call, so we need external input from the node operator.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala [hunk 13]
```diff
@@ -18,25 +18,32 @@ package fr.acinq.eclair.blockchain.fee
...
-  def getMempoolMinFeeratePerKw(): FeeratePerKw
-  // @formatter:on
+// @formatter:off
+sealed trait ConfirmationPriority {
+  def getFeerate(feerates: FeeratesPerKw): FeeratePerKw = this match {
+    case ConfirmationPriority.Slow => feerates.slow
+    case ConfirmationPriority.Medium => feerates.medium
+    case ConfirmationPriority.Fast => feerates.fast
+  }
+override def toString: String = super.toString.toLowerCase
```
[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/OnChainFeeConf.scala]: Done in dbf895211.

File: eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalStateSpec.scala (hunk 10 above)
Comment (pm47) [reply eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalStateSpec.scala]: Done in dbf895211.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala (hunk 12 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala]: Done in dbf895211.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala (hunk 11 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala:247]: I thought about that, but I figured that in this scenario eclair would have been restarted, so there wouldn't be a replacement.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala:247]: That's true, good point!

**Comment by pm47** [comment]
Rebased.

File: eclair-core/src/main/resources/reference.conf [hunk 14]
```diff
@@ -212,20 +212,10 @@ eclair {
...
-      // target for the commitment transaction when we have no htlcs to claim (used in force-close scenario) *do not change this unless you know what you are doing*
-      commitment-without-htlcs = 12
-      // target for the mutual close transaction
-      mutual-close = 12
-      // target for the claim main transaction (tx that spends main channel output back to wallet)
-      claim-main = 12
-      // when our utxos count is below this threshold, we will use more aggressive confirmation targets in force-close scenarios
-      safe-utxos-threshold = 10
+    // confirmation priority for each transaction type, can be slow/medium/fast
+    confirmation-priority {
```
[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/resources/reference.conf:213]: Done in [5d648bf](https://github.com/ACINQ/eclair/pull/2696/commits/5d648bfb66e359b558a711930d3770e26bccd45b).

File: eclair-core/src/main/resources/reference.conf [hunk 15]
```diff
@@ -201,31 +201,18 @@ eclair {
...
-    default-feerates { // those are per target block, in satoshis per kilobyte
-      1 = 210000
-      2 = 180000
-      6 = 150000
-      12 = 110000
-      36 = 50000
-      72 = 20000
-      144 = 15000
-      1008 = 5000
+    default-feerates { // those are per target block, in satoshis per byte
```
Comment (t-bast) [inline eclair-core/src/main/resources/reference.conf:204]:
```suggestion
    default-feerates { // the following values are in satoshis per byte
```

File: docs/release-notes/eclair-vnext.md [hunk 16]
```diff
@@ -4,7 +4,23 @@
...
+
+Default setting is `medium` for both funding and closing. Node operators may configure their values like so:
+```eclair.conf
+on-chain-fees {
+    confirmation-priority {
+      funding = fast
+      closing = slow
+    }
+}
+```
```
Comment (t-bast) [inline docs/release-notes/eclair-vnext.md:24]:
```suggestion
This configuration section replaces the previous `eclair.on-chain-fees.target-blocks` section.

```

File: docs/release-notes/eclair-vnext.md [hunk 17]
```diff
@@ -4,7 +4,23 @@
...
+the normal closing priority is used.
+
+Default setting is `medium` for both funding and closing. Node operators may configure their values like so:
+```eclair.conf
+on-chain-fees {
+    confirmation-priority {
+      funding = fast
+      closing = slow
+    }
+}
```
Comment (t-bast) [inline docs/release-notes/eclair-vnext.md:22]: If node operators directly copy-paste this in their `eclair.conf`, it won't work, it must be wrapped in an `eclair` section:

```suggestion
eclair.on-chain-fees.confirmation-priority {
    funding = fast
    closing = slow
}
```

File: eclair-core/src/main/resources/reference.conf (hunk 15 above)
Comment (pm47) [reply eclair-core/src/main/resources/reference.conf]: Done in 22160f28e.

File: docs/release-notes/eclair-vnext.md (hunk 17 above)
Comment (pm47) [reply docs/release-notes/eclair-vnext.md]: Done in 22160f28e.

**Comment by pm47** [comment]
>>Not sure exactly how to proceed though!

>I don't know what we can do honestly, since we want to handle bitcoind restarts...I don't think we want to persist the latest fees, it's better to avoid a frequent DB call, so we need external input from the node operator (which is why hard-coded values in eclair.conf makes sense, even though it's not very satisfying).

Let's defer this to a later time.

**Review by t-bast — APPROVED** [review]
